### ISSUE TITLE
transmitter fails after several days

### ISSUE THREAD

**Issue by OlafTheSnowman** [jgeisler0303/can2mqtt#3]
Hello jgeisler0303,

after several days the transmitter (mqtt -> can) fails and does not leave any hint within the logs (debug mode). The receiver part is still working fine. Stop/start the application. solves the problem

What do you think could be the problem?

Thank you

Olaf

**Comment by jgeisler0303** [comment]
Hi,
unfortunately there are at least two third party libraries involved, so debugging this my be not so easy.

But the first thing I would do is try to find out, where the messages get lost. You could e.g. add more logging at the beginning of ``def on_message(client, userdata, message):`` to see if the mqtt messages are received. If not then maybe your mqtt server forgets the subscriptions (or maybe this is even intended behavior and the subscriptions must be renewed every now and then?).

Maybe you could query your mqtt server for active subscriptions to find further narrow down this possibility and rule out that the mqtt python library drops the messages before they can get passed on to ``on_message``.

If the messages arrive in ``on_message`` but don't get sent on the CAN bus, maybe the CAN library crashes. To further prove or disprove this, you could configure a periodic CAN message like a sync to see it that continues after the mqtt messages stop.

Also you could add some logging before the line ``m= can.Message(extended_id= False, arbitration_id= canid, data= data)`` if CAN messages are tried to be sent.

I know this is very frustrating to have to wait for an error like this for so long. So you should probably add ample instrumentation (logging) in the hope that you catch the part that fails.

Good luck and please leave comment, when you find the solution.

**Comment by OlafTheSnowman** [comment]
Hi, OK I added the additional logs. Let's what happens... I will report about the progress, may be this takes a few days. 
Thank you

**Comment by OlafTheSnowman** [comment]
Hi jgeisler0303,

mosquitto reports:

1622694479: Client can2mqtt has exceeded timeout, disconnecting.
1622694480: Socket error on client can2mqtt, disconnecting.

and can2mqtt throws:

2021-06-03 06:20:55,990 ERROR:Error publishing message "/CANOpen/ID2/SDO6200" to topic "4F,00,62,04,00,00,00,00". Return code 4: The client is not currently connected.

It seems to me that mosquitto drops the client "can2mqtt" after some time!? 

Do you think it is possible to integrate a "reconnect" functionality for the mqtt client?

Have a nice day

**Comment by jgeisler0303** [comment]
Hi Olaf,
thanks for the investigation. Its good news because now we know the root cause. Its a bit strange that this happens because the ``loop_start`` in line 347 should [handle reconnection](https://pypi.org/project/paho-mqtt/#connect-reconnect-disconnect), but who knows...

Would you like to try to fix this yourself, fork the repo and send me a pull request or just send me the fixed code?

I would probably just try to fix this quick and dirty:
1. Change line 408 to a warning
2. Immediately below line 408 and indented appropriately add a line ``client.reconnect()``
3. Below that paste a copy of lines 406 thru 408 appropriately indented and with the original error message of line 408
4. (Maybe wrap all this into an if statement that checks if the error code in ``r[0]`` is 4)

For testing this, maybe it is possible to force the mqtt host to disconnect the can2mqtt client, maybe just by restarting the host. This way you don't have to wait for such along time again for the error to occur.

Good Luck

**Comment by OlafTheSnowman** [comment]
Hello Jens,

I added the code modifications as you mentioned, see below... A restart of the mosquitto service doesn't cause the expected exception, if you restart mosquitto the can2mqtt app does not react any longer. May be there is an other hint if the reconnect problem occurs while normal operation, lets see what happens.

Thank you for your support

                            if not (r[0] == mqtt.MQTT_ERR_SUCCESS):
                                logging.info("MQTT error")
                                logging.warning("Error publishing message \"%s\" to topic \"%s\". Return code %s: %s" % (t, p, str(r[0]), mqtt.error_string(r[0])))
                                if (r[0] == 4):
                                    logging.info("MQTT reconnect")
                                    client.reconnect()
                                    r= client.publish(t, p)
                                    if not (r[0] == mqtt.MQTT_ERR_SUCCESS):
                                        logging.error("Error publishing message \"%s\" to topic \"%s\". Return code %s: %s" % (t, p, str(r[0]), mqtt.error_string(r[0]))

**Comment by OlafTheSnowman** [comment]
Hello Jens,

I decided for some reasons to switch from RPi3 to a PC with much better performance. I had already a longer time the suspicion, that the performance of an RPi3 is not enough for OpenHAB, mosquitto and can2mqtt. The new system is now up for several days and it is running much better, especially the execution of OpenHAB rules.

So I will close this issue. Thank you for your great support.

**Comment by jgeisler0303** [comment]
Hi Olaf,
thanks again for your work and sharing your insights. I hope the strange behavior was due to your RPi3 being temporarily overloaded and is gone now with a more capable PC.

Maybe your previous post is a hint of what was going on. Maybe a reconnect doesn't suffice when the connections is forcefully interrupted. I guess I will make the change you posted but instead of a reconnect I will introduce a full new connect.

Cheers, Jens

**Comment by OlafTheSnowman** [comment]
Hi Jens,
yes the system is working much better now, I also sorted out another timing problem. Over the time I'm using can2mqtt I also added a support to run this script as a service with a start/stop script, so it could be run automaticly at system startup.

Did you ever spend a thought on port the script to python3? I tried this but one of the message translations expects an other data type.

Have a nice day, Olaf
